nomad/plan_apply.go

```diff
@@ -159,6 +159,7 @@ func (p *planner) planApply() {
 			if err != nil {
 				p.logger.Error("failed to update snapshot state", "error", err)
 				pending.respond(nil, err)
+				planIndexCh = nil
```

Comment:
> Should this be on L157 right after we receive from the chan?

Yup, good catch. I was focused on the error branch because there's where the bug showed up, but there's no reason to keep it once we've received a value.

> To put another way: I think `planIndexCh` is always a one-shot chan. It will always receive exactly one index value and never get reused.

Yes, agreed!

> If you think that's accurate I wonder if we should wrap planIndexCh in a little struct with a method like:
> ...
> We also use `planIndexCh == nil` to signal "no outstanding plan application", so our struct would probably need a helper method for that too. Unsure if the extra struct would simplify or complicate reading this code... just an idea.

Hm... I like this idea in principle. But the logic here leans heavily on the specific semantics of channels (ex. a closed channel isn't nil) and wrapping it in a struct just seems like it'll obscure that by moving the logic away from where we're using it. The fallthrough `select` on line 104-113 makes this especially gross because it relies on the fact that we get 0 on error, so we can't even call `recv()` and check for 0 there so it needs it's own `index, ok := recvOrDefault()` function.

<details><summary>some example code</summary>

```go

type asyncPlanIndex struct {
	planIndexCh chan uint64
}

func (a *asyncPlanIndex) hasOutstandingPlan() bool {
	return a.planIndexCh == nil
}

func (a *asyncPlanIndex) recv() uint64 {
	v := <-a.planIndexCh
	a.planIndexCh = nil
	return v
}

func (a *asyncPlanIndex) recvOrDefault() (uint64, bool) {
	select {
	case v := <-a.planIndexCh:
		return v, true
	default:
		return 0, false
	}
}

func (a *asyncPlanIndex) send(v uint64) {
	a.planIndexCh <- v
	close(a.planIndexCh)
}

```

</details>

Reply:
> But the logic here leans heavily on the specific semantics of channels

Oof yeah, this is a pattern to watch out for in the future. Not always a problem but clearly made this code more fragile than if a little effort had been put into encapsulating the logic in a little purpose-built struct.